codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java

```diff
@@ -61,27 +68,92 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
                 Http2Headers http2Headers = HttpUtil.toHttp2Headers(httpMsg);
                 Http2ConnectionEncoder encoder = encoder();
 
-                if (hasData) {
-                    promiseAggregator = new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
-                    encoder.writeHeaders(ctx, streamId, http2Headers, 0, false, promiseAggregator.newPromise());
-                    httpMsgNeedRelease = false;
-                    encoder.writeData(ctx, streamId, httpMsg.content(), 0, true, promiseAggregator.newPromise());
-                    promiseAggregator.doneAllocatingPromises();
+                if (httpMsg instanceof FullHttpMessage) {
+                    FullHttpMessage fullHttpMsg = (FullHttpMessage) httpMsg;
+                    boolean hasData = fullHttpMsg.content().isReadable();
+                    Http2Headers trailingHeaders = fullHttpMsg.trailingHeaders().isEmpty() ? null :
+                            HttpUtil.toHttp2Headers((LastHttpContent) fullHttpMsg);
+                    if (hasData) {
+                        promiseAggregator = new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+                        encoder.writeHeaders(ctx, currentStreamId, http2Headers, 0, false,
+                                promiseAggregator.newPromise());
+
+                        final boolean hasTrailingHeaders = trailingHeaders != null && !trailingHeaders.isEmpty();
+                        encoder.writeData(ctx, currentStreamId, fullHttpMsg.content(), 0, !hasTrailingHeaders,
+                                promiseAggregator.newPromise());
+                        release = false;
+
+                        if (hasTrailingHeaders) {
+                            encoder.writeHeaders(ctx, currentStreamId, trailingHeaders, 0, true,
+                                    promiseAggregator.newPromise());
+                        }
+
+                        promiseAggregator.doneAllocatingPromises();
+                    } else {
+                        encoder.writeHeaders(ctx, currentStreamId, http2Headers, 0, true, promise);
+                    }
+                } else {
+                    encoder.writeHeaders(ctx, currentStreamId, http2Headers, 0, false, promise);
+                }
+            } catch (Throwable t) {
+                if (promiseAggregator == null) {
+                    promise.tryFailure(t);
+                } else {
+                    promiseAggregator.setFailure(t);
+                }
+            } finally {
+                if (release) {
+                    ReferenceCountUtil.release(httpMsg);
+                }
+            }
+        } else if (msg instanceof HttpContent) {
+            final HttpContent httpContent = (HttpContent) msg;
+            boolean release = true;
+            SimpleChannelPromiseAggregator promiseAggregator = null;
+            try {
+                if (currentStreamId == 0) {
```

Comment:
I'm not sure we should check this here (or worry about resetting `currentStreamId = 0`).  The HTTP/2 codec is responsible for enforcing this.